### ISSUE TITLE
Scriptables' context is missing elements and options

### ISSUE THREAD

**Issue by joshkel** [chartjs/chartjs-plugin-annotation#375]
I'm using a scriptable `display` property to control when annotations can be shown, based on their options. (This is the same use case I'd originally asked about in #244.) This was working in an earlier copy of chartjs-plugin-annotation (circa December 2020), but in the current 1.0.0-rc, it's giving problems:

* The first time `display` is called, the context parameter's `element` member is completely missing.
* The second time it's called, `element` is present, but it has no `options` member.

If `display` is called again (for example, because the chart updates from showing or hiding a dataset), then `element.options` is correctly populated.

See https://codesandbox.io/s/chartjs-annotation-context-c95k6 for a demo. The console will show error messages when the page is first loaded. Clicking a dataset's legend item to show or hide the dataset will cause the annotations to be correctly shown or hidden.

**Comment by stockiNail** [comment]
@joshkel I think the issue is here:

https://github.com/chartjs/chartjs-plugin-annotation/blob/b9bc2955d20607d9b6d4e336b2fe34dde153a326/src/annotation.js#L65

The `display` property is used and being in a proxy options, the callback is invoked by `chart` context type.

Add `console.log(context.type);` in your callback and you can see the type == `chart` instead of `annotation` (the first call).

@kurkle I think it's a bug. The elements are created in the `afterUpdate` hook which is invoked after  `afterDataLimits` where the options `display` is used. Before the afterUdate we must be sure do not invoke any options to avoid this issue. Make sense?

I think it sounds related to this issue and the behavior of the proxy options: https://github.com/chartjs/Chart.js/issues/8623

**Comment by joshkel** [comment]
Thanks, @stockiNail . I tracked down the issue as far as that but realized I didn't understand Chart.js's proxied context implementation well enough (is there documentation on it?), so I wasn't confident of a fix.

Would it work to create (mostly empty?) elements objects and initialize their contexts sooner (e.g., from `beforeUpdate`, when the annotations array is first set)?

#300 was a similar issue from before the proxy options implementation.

**Comment by stockiNail** [comment]
> (is there documentation on it?),

I'm not sure about documentation because I have studied it (from the code) for my lib because I didn't recognize at the beginning its behaviors.

> Would it work to create (mostly empty?) elements objects and initialize their contexts sooner (e.g., from `beforeUpdate`, when the annotations array is first set)?
> 
> #300 was a similar issue from before the proxy options implementation.

I think your proposal didn't solve completely this issue.

The issue is not related to the missing element, but the callback is invoked with a `chart` context.

The `resolveAnnotationOptions(resolver)` function should be executed in `beforeUpdate` otherwise the callback assigned to `display` will be invoked always with a `chart` context, instead of `annotation` one (proxy behavior).

Nevertheless the workaround is to add a check in your callback to test if the context you are receiving as argument is consistent (annotation type):

```javascript
const display = (context) => {
  if (context.type !== 'annotation') {
    return true; // checks the scales for all annotation but it should work (I didn't test)
  }
  // your logic here when context is annotation type
};
```

About the missing options, the callbacks are invoked before the options are added to element (the first call). Maybe that part of code could be reviewed.

**Comment by stockiNail** [comment]
I think there are other statements in the code where this issue could occur (where an option is invoked before `resolveAnnotationOptions` function and could be scriptable):

```javascript
function getScaleLimits(scale, annotations) {
  const axis = scale.axis;
  const scaleID = scale.id;
  const scaleIDOption = axis + 'ScaleID';
  let min = valueOrDefault(scale.min, Number.NEGATIVE_INFINITY);
  let max = valueOrDefault(scale.max, Number.POSITIVE_INFINITY);
  for (const annotation of annotations) {
    if (annotation.scaleID === scaleID) {    // <-- if 'scaleID' is a scriptable options
      for (const prop of ['value', 'endValue']) { 
        const raw = annotation[prop];    // <-- if 'value' and/or 'endValue' are scriptable options
        if (raw) {
          const value = scale.parse(raw);
          min = Math.min(min, value);
          max = Math.max(max, value);
        }
      }
    } else if (annotation[scaleIDOption] === scaleID) {  // <-- if the axis+'scaleID' is a scriptable options
      for (const prop of [axis + 'Min', axis + 'Max', axis + 'Value']) {
        const raw = annotation[prop];  // <-- if the axis+'Min', axis+'Max' and/or axis+'Value' are scriptable options
        if (raw) {
          const value = scale.parse(raw);
          min = Math.min(min, value);
          max = Math.max(max, value);
        }
      }
    }
  }
  return {min, max};
}
```

**Comment by kurkle** [comment]
`display:(context,options) => ...`

**Comment by kurkle** [comment]
Also the `display` property was missed in docs:
https://www.chartjs.org/chartjs-plugin-annotation/guide/options.html#option-context

**Comment by joshkel** [comment]
Thanks, @kurkle.

So, is the current behavior desired? If so, the TypeScript type definitions should be changed?

```
// Some scriptable options may be invoked before the element is fully initialized.
export interface PartialEventContext {
	chart: Chart,
	element?: Partial<AnnotationElement>,
}

export interface CoreAnnotationOptions extends AnnotationEvents {
  display?: boolean | ((context: PartialEventContext, options: AnnotationOptions) => boolean);
```

I can submit a PR for this if you like.

I noticed another type-related issue while investigating this issue: `context.element.options.type` is one of the `...Annotation.id` values (e.g., `boxAnnotation`), while `options.type` is an `annotationTypes` / `AnnotationTypeRegistry` key (e.g., `box`). However, the TypeScript type definitions say that it's always an `AnnotationTypeRegistry` value. I can open a separate issue for this or try and open a PR for this as well if you like.

**Comment by kurkle** [comment]
Not sure if its desired. Its a compromise. Earlier, all of the element options were resolved early, before the scales were laid out etc, so you could not get a tick value for example for annotation value. So its desirable to resolve the options that can be resolver later, when then scales are laid out etc.

But then the adjustScaleRanges need to be done before the scales are laid out.. 

I think the current way covers more use cases, although adding the confusion to options being resolved with different contexts.
Another option would be to create the elements early, but not resolve the options. But I'm not sure an empty element with no valid properties would be useful.

So.. the behavior is there, and the types are not aligned, so I'd welcome a PR for those. 

The type issue could be a bug.

**Comment by joshkel** [comment]
Thanks.

I updated the type definitions and added some comments to the docs to try and explain these discussions. Feedback welcome.

**Comment by kurkle** [comment]
> I can open a separate issue for this or try and open a PR for this as well if you like.

Separate issue would be good. This can be closed after that?

**Comment by joshkel** [comment]
I created #388 to track the `options.type` discrepancy. I apologize for the delay in getting back to this. Thanks for your help!
